### PR TITLE
Subject merge

### PR DESCRIPTION
**Pull request by smangelsdorf**
Merges an invited subject into an existing one, if the existing subject accepts the invitation.

Fixes the scenarios described in #35 and #36

### CODE REVIEW

File: app/models/subject.rb [hunk 1]
```diff
@@ -22,12 +22,57 @@ def accept(invitation, attrs)
...
+      merge(invited_subject) if invited_subject != self
+    end
+  end
+
+  def merge(other)
+    transaction do
+      merge_roles(other)
+      merge_attributes(other)
+
+      other.reload # HACK: Why is this needed? Tests fail without it.
```
Comment (rianniello) [inline app/models/subject.rb]: Interesting... assuming the reload doesn't work as expected in the test either?

Comment (smangelsdorf) [inline app/models/subject.rb]: It's related to the fact that `merge_roles` and `merge_attributes` get `other` into a weird state when they're performing deletions.

The interaction between that and the `audited` gem was causing a failure when trying to save the audit record.

As suggested by the code comment, I'm happy to fix this properly if you have any ideas.

Comment (smangelsdorf) [inline app/models/subject.rb]: I had an idea, and it seems to have fixed it. It was caused by explicitly deleting child objects that would later be automatically deleted via the ActiveRecord association. The `audited` gem wasn't build to handle that edge case.

File: lib/authentication/subject_receiver.rb [hunk 2]
```diff
@@ -47,5 +49,23 @@ def update_subject(subject, attrs)
...
+    def subject_scope(attrs)
+      t = Subject.arel_table
+      Subject.where(t[:targeted_id].eq(attrs[:targeted_id])
+        .or(t[:shared_token].eq(attrs[:shared_token])))
+    end
+
+    def check_subject(subject, attrs)
+      require_nil_or_equal(subject.targeted_id, attrs[:targeted_id])
+      require_nil_or_equal(subject.shared_token, attrs[:shared_token])
+      subject
```
Comment (rianniello) [inline lib/authentication/subject_receiver.rb]: Does `subject` need to be returned? Doesn't look like it's used.

Comment (smangelsdorf) [inline lib/authentication/subject_receiver.rb]: Get that into ya.

Comment (rianniello) [inline lib/authentication/subject_receiver.rb]: Thanks, my job here is done.

image

**Pull request merged**